src/python/WMCore/MicroService/MSPileup/MSPileupData.py

```diff
@@ -0,0 +1,204 @@
+        # python 2/3 super style
+        # super(MSPileupData, self).__init__(msConfig, **kwargs)
+        # python 3 super style
+        super().__init__(msConfig, **kwargs)
+
+        self.msConfig.setdefault("mongoDBRetryCount", 3)
+        self.msConfig.setdefault("mongoDBReplicaSet", None)
+        self.msConfig.setdefault("mongoDBPort", None)
+        self.msConfig.setdefault("mockMongoDB", False)
+
+        # A full set of valid database connection parameters can be found at:
+        # https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html
+        mongoIndex = IndexModel('pileupName', unique=True)
+        collection = self.msConfig.get('mongoDBCollection', 'msPileupCollection')
+        msOutDBConfig = {
+            'database': self.msConfig['mongoDB'],
+            'server': self.msConfig['mongoDBServer'],
+            'replicaSet': self.msConfig['mongoDBReplicaSet'],
+            'port': self.msConfig['mongoDBPort'],
+            'username': self.msConfig['mongoDBUser'],
+            'password': self.msConfig['mongoDBPassword'],
+            'mockMongoDB': self.msConfig['mockMongoDB'],
+            'connect': True,
+            'directConnection': False,
+            'logger': self.logger,
+            'create': True,
+            'collections': [(collection, mongoIndex)]}
+        mongoDB = MongoDB(**msOutDBConfig)
+        self.msDB = getattr(mongoDB, self.msConfig['mongoDB'])
+        self.dbColl = self.msDB[collection]
+
+    def create(self, pdict):
+        """
+        Create and return pileup data from campaigns dictionary
+
+        :param pdict: a dictionary representing MSPileup data
+        :return: list of MSPileupError or empty list
+        """
+        # first, create MSPileupObj which will be validated against its schema
+        try:
+            obj = MSPileupObj(pdict)
+            doc = obj.data
+        except Exception as exp:  # pylint: disable=broad-except
+            msg = f"Failed to create MSPileupObj, {exp}"
+            self.logger.exception(msg)
+            err = MSPileupError(pdict, msg)
+            return [err.DuplicateDocument()]
+
+        # insert document into underlying DB
+        try:
+            self.dbColl.insert_one(doc)
+        except errors.DuplicateKeyError:
+            msg = f"Failed to insert: {doc}, it already exist in DB"
+            self.logger.exception(msg)
+            err = MSPileupError(doc, msg)
+            return [err.DuplicateDocument()]
+        return []
+
+    def update(self, doc):
+        """
+        Update pileup data with provided input
+
+        :param doc: a dictionary of pieleup data to be updated
+        :return: list of MSPileupError or empty list
+        """
+        # check if given document contains pileup Name (unique key)
+        pname = doc.get('pileupName', '')
+        if not pname:
+            err = MSPileupError(doc, f'No document found for {pname}')
+            return [err.NoKeyFound()]
```

Comment:
I know we discussed this on Monday and it was the agreement for propagating an error to the end user. However, if we implement it at this layer, it will increase the complexity of the MSPileup daemon, because the same list object returned by this method can either contain an error object, or the pileup object itself.

I wonder if this logic should be implemented in the layer above, aka the MSPileup daemon/app running the whole logic of the service. For instance:
1. read data from the database
2. execute the logic of data placement and check
3. update the pileup object
4. persist the pileup object in the database